pkg/agent/flexvolume.go

```diff
@@ -0,0 +1,243 @@
+	}
+
+	logger.Infof("Creating Volume attach Resource %s/%s", clusterName, crdName)
+
+	volumeattachObj := &volumeattach.Volumeattach{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      crdName,
+			Namespace: clusterName,
+			Labels: map[string]string{
+				volumeattachLabelNodeKey: attachOpts.NodeName,
+			},
+		},
+		Spec: volumeattach.VolumeattachSpec{
+			VolumeID:    attachOpts.Image,
+			VolumeGroup: attachOpts.Pool,
+			Node:        attachOpts.NodeName,
+		},
+		Status: volumeattach.VolumeattachStatus{
+			State:   volumeattach.VolumeattachStatePending,
+			Message: volumeattachPendingMessage,
+		},
+	}
+	volumeattachObj.APIVersion = fmt.Sprintf("%s/%s", k8sutil.CustomResourceGroup, k8sutil.V1Alpha1)
+	volumeattachObj.Kind = reflect.TypeOf(volumeattach.Volumeattach{}).Name()
+
+	var result volumeattach.Volumeattach
+	body, _ := json.Marshal(volumeattachObj)
+	uri := fmt.Sprintf("apis/%s/%s/namespaces/%s/%s", k8sutil.CustomResourceGroup, k8sutil.V1Alpha1, clusterName, volumeattach.CustomResourceNamePlural)
+
+	err = c.clientset.Core().RESTClient().Post().
+		RequestURI(uri).
+		Body(body).
+		Do().Into(&result)
+	if err != nil {
+		if !kserror.IsAlreadyExists(err) {
+			return fmt.Errorf("Rook: Failed to create Volumeattach CRD: %+v", err)
+		}
+
+		// Volumeattach resource exists.
+		logger.Infof("Volume attach Resource %s/%s already exists", clusterName, crdName)
+		volumeattachResource, err := getVolumeattachResource(c.clientset, clusterName, crdName)
+		if err != nil {
+			return fmt.Errorf("Failed to get Volumeattach resource %s/%s: %+v", clusterName, crdName, err)
+		}
+
+		// If volume attach failed previously, change state back to Pending to retry.
+		if volumeattachResource.Status.State == volumeattach.VolumeattachStateFailed {
+			volumeattachResource.Kind = reflect.TypeOf(volumeattach.Volumeattach{}).Name()
+			volumeattachResource.APIVersion = fmt.Sprintf("%s/%s", k8sutil.CustomResourceGroup, k8sutil.V1Alpha1)
+			volumeattachResource.Status.State = volumeattach.VolumeattachStatePending
+			volumeattachResource.Status.Message = volumeattachPendingMessage
+			volumeattachResource.UpdateCRD(c.clientset)
+		}
+	}
+
+	reply.Status = "Success"
+	reply.DevicePath = crdName
+
+	return nil
+}
+
+// Detach a rook volume to the node
+func (c *FlexvolumeController) Detach(detachOpts *AttachOptions, reply *Result) error {
+
+	// the CRD name is the same as image id, which is the form of clusterName-pvc.
+	// Parsing namespace from the crdName since Kubernetes flexvolume doesnt pass the volume source
+	crdName := detachOpts.Image
+	namespace := crdName[0:strings.LastIndex(crdName, "-pvc")]
+
+	logger.Infof("Deleting Volume attach Resource %s/%s", namespace, crdName)
+	uri := fmt.Sprintf("apis/%s/%s/namespaces/%s/%s", k8sutil.CustomResourceGroup, k8sutil.V1Alpha1, namespace, volumeattach.CustomResourceNamePlural)
+	err := c.clientset.Core().RESTClient().Delete().
```

Comment:
Is there any retry for failed detaches?  When we delete the volume attach CRD, the remote agent watching for CRD events will get the delete event and attempt to delete.  If that fails, the CRD is deleted, the remote agent won't get any other events, and it'll never retry again.

Or does detach retry happen some other way?

Reply:
you are right. We need to have code to keep the sync between CRDs and volume attachment.  But now that we are not using the attacher-detacher controller, we could react when there is a failure